src/mo_frontend/typing.ml

```diff
@@ -1048,11 +1051,49 @@ and infer_exp'' env exp : T.typ =
       else env
     in
     infer_obj env' obj_sort.it dec_fields exp.at
-  | ObjE exp_fields ->
+  | ObjE (exp_bases, exp_fields) ->
+    let open List in
     check_ids env "object" "field"
-      (List.map (fun (ef : exp_field) -> ef.it.id) exp_fields);
-    let fts = List.map (infer_exp_field env) exp_fields in
-    T.Obj (T.Object, List.sort T.compare_field fts)
+      (map (fun (ef : exp_field) -> ef.it.id) exp_fields);
+    let fts = map (infer_exp_field env) exp_fields in
+    let fls = map (fun {T.lab; _} -> lab) fts in
+    let bases = map (fun b -> infer_exp env b, b) exp_bases in
```

Comment:
```suggestion
    let bases = map (fun b -> infer_exp_promote env b, b) exp_bases in
```
This should promote bases of varialbe types to their bounds, letting you merge a base expression whose type is generic, but bounded by a record. The typeing code for DotE does similar.

Reply:
This improved things: a973e4220